### ISSUE TITLE
CC0057 (VB) suggests removing parameter but it is required by a matching delegate

### ISSUE THREAD

**Issue by kwhitefoot** [code-cracker/code-cracker#791]
#Bug

This diagnostic correctly finds that a parameter is unused in the function where it is declared but it fails to check that the signature is required by a matching delegate or variable declaration.

In cases where the function is assigned to a variable the declaration of the variable should be checked and the fix should not be offered if the variable declaration requires the unused parameter.

Alternatively it would be neat if the diagnostic could search for all of the functions that are used this way and check if the parameter is unused in all of them.  Then it could remove them from both the function signatures and the variable declaration.  

````vb.net

  Private x As Func(Of Boolean, Double) = AddressOf x0

  Private Function x0(b As Boolean) As Double
    Return 0.0
  End Function
````

Current output after fix applied (if it is a code fix bug):

````vb.net

  Private x As Func(Of Boolean, Double) = AddressOf x0

  Private Function x0() As Double
    Return 0.0
  End Function

**Comment by wmjordan** [comment]
This code fix can break the code. It also exists in the C# component.
Fixing this bug might harm the performance terribly, since the analyzer has to search the whole solution to find out whether the function is referenced as a delegate.

**Comment by giggio** [comment]
Bug confirmed.

**Comment by giggio** [comment]
@wmjordan yes, you are right, we will not be able to detect the references during analysis, we would have to do it during the execution of the fix. So, should we accept the false positive, and, in the fix, do nothing?
This is one of the most useful analyzers we got, I don't want to disable it.
@code-cracker/core @code-cracker/collaborators I would like your opinions too.

**Comment by wmjordan** [comment]
@giggio, I inspected my cases and the code sample provided by @kwhitefoot. I found something in common. Both of us are using **private** methods as `delegate`s. You may limit the analyzer work within the scope of the current class. If the method is in a `public`/`internal` nested class, examine the outer class too.
It helps suppressing those false positive reports.

**Comment by giggio** [comment]
@wmjordan It is a possibility, but this would mean that we can only run this analysis on private members...

**Comment by thomaslevesque** [comment]
> but this would mean that we can only run this analysis on private members...

For what it's worth, that's what ReSharper does, and I think it makes sense.
